sequence_processing_pipeline/Pipeline.py

```diff
@@ -292,14 +317,220 @@ def generate_sample_information_files(self):
+        # defined in both cases.
         results = []
 
-        for result in bioinformatics.to_dict('records'):
-            results.append({'project_name': result['Sample_Project'],
-                            'qiita_id': result['QiitaID']})
+        if self.mapping_file is not None:
+            sample_project_map = {}
+            for sample, project in zip(self.mapping_file.sample_name,
+                                       self.mapping_file.project_name):
+                if project not in sample_project_map:
+                    sample_project_map[project] = []
+                sample_project_map[project].append(sample)
+
+            # TODO: Replace 'something' w/proper Qiita ID
+            for project in sample_project_map:
+                qiita_id = project.split('_')[-1]
+                results.append(
+                    # assume project names end in qiita ids
+                    {'project_name': project, 'qiita_id': qiita_id})
+        else:
+            bioinformatics = self.sample_sheet.Bioinformatics
+            for result in bioinformatics.to_dict('records'):
+                results.append({'project_name': result['Sample_Project'],
+                                'qiita_id': result['QiitaID']})
 
         return results
+
+    @staticmethod
+    def is_mapping_file(mapping_file_path):
+        try:
+            Pipeline._validate_mapping_file(mapping_file_path)
+            return True
+        except PipelineError:
+            return False
+
+    @staticmethod
+    def _validate_mapping_file(mapping_file_path):
+        exp = {'barcode', 'library_construction_protocol', 'mastermix_lot',
+               'sample_plate', 'center_project_name', 'instrument_model',
+               'tm1000_8_tool', 'well_id', 'tm50_8_tool', 'well_description',
+               'run_prefix', 'run_date', 'center_name', 'tm300_8_tool',
+               'extraction_robot', 'experiment_design_description',
+               'platform', 'water_lot', 'project_name', 'pcr_primers',
+               'sequencing_meth', 'plating', 'orig_name', 'linker', 'runid',
+               'target_subfragment', 'primer', 'primer_plate', 'sample_name',
+               'run_center', 'primer_date', 'target_gene', 'processing_robot',
+               'extractionkit_lot'}
+        try:
+            df = pd.read_csv(mapping_file_path, delimiter='\t')
+
+            # if the two sets of headers are equal, then return the dataframe
+            # and no error message.
+            if len(exp - set(df.columns)) == 0:
+                return df
+
+            msg = 'missing columns: %s' % ', '.join(exp - set(df.columns))
+        except pd.errors.ParserError:
+            # ignore parser errors as they obviously prove this is not a
+            # valid mapping file.
+            msg = 'could not parse file "%s"' % mapping_file_path
+
+        raise PipelineError(msg)
+
+    def _generate_dummy_sample_sheet(self, first_read, last_read,
```

Comment:
Just to confirm here: will this use the project_name in the mapping file? Asking because is important that the resulting read files (R1, R2, Index) use that name for that project.

Reply:
What bcl-convert does is generate fastq files with file-names based on sample-id. Since we only define one sample in the sheet, it appears that it's just been an arbitrary name until now.

Right now what I'm doing is naming the sample for the run-id, and then bcl-convert is run once and the results are copied into multiple project folders according to the projects found in the mapping file. The three/four read files are all under the right project folder(s), but they retain the filename beginning with the run-id.

If you want I could rename them as I'm copying them, but I would do that outside of this function and where I'm doing the copying.

Reply:
AFAIK the sample_sheet will use as sample_name something similar to the project_name so the files have that sample name. IMOO is fine to use whatever as long as the final name (when is mv to the upload folder) has the [project_name]_[run_id] as filename prefixes. Note that this should also be the run_prefix for all samples in the prep info file.

Reply:
This issue has been resolved. Confirmed run-id as a prefix is fine for the fastq files so long as they're in folders with the proper project names.